modules/service/dataset-pruning-service.js

```diff
@@ -209,6 +216,135 @@ class DatasetPruningService {
+     *         chosen: false,
+     *         foundOffer: false,
+     *         foundPurchase: false,
+     *     }
+     * }
+     * @param datasets
+     * @returns {{}}
+     */
+    repackLowEstimatedValueDatasets(datasets) {
+        const repackedDatasets = {};
+        datasets.forEach((dataset) => {
+            if (!repackedDatasets[dataset.data_set_id]) {
+                repackedDatasets[dataset.data_set_id] = {
+                    dataInfos: [],
+                    bids: [],
+                };
+            }
+            if (!repackedDatasets[dataset.data_set_id].expiry
+                || repackedDatasets[dataset.data_set_id].expiry < dataset.expiry) {
+                repackedDatasets[dataset.data_set_id].expiry = dataset.expiry;
+            }
+            const foundDataInfoId = repackedDatasets[dataset.data_set_id].dataInfos
+                .includes(dataset.data_info_id);
+            if (!foundDataInfoId) {
+                repackedDatasets[dataset.data_set_id].dataInfos.push(dataset.data_info_id);
+            }
+
+            if (dataset.bid_id) {
+                const foundBidId = repackedDatasets[dataset.data_set_id].bids
+                    .includes(dataset.bid_id);
+                if (!foundBidId) {
+                    repackedDatasets[dataset.data_set_id].bids.push(dataset.bid_id);
+                    if (dataset.bid_status === 'CHOSEN') {
+                        repackedDatasets[dataset.data_set_id].chosen = true;
+                    }
+                }
+            }
+            if (dataset.offer_id) {
+                repackedDatasets[dataset.data_set_id].foundOffer = true;
+            }
+            if (dataset.purchase_id) {
+                repackedDatasets[dataset.data_set_id].foundPurchase = true;
+            }
+        });
+        return repackedDatasets;
+    }
+
+    getLowEstimatedValueIdsForPruning(repackedDatasets) {
+        const idsForPruning = {
+            datasetsToBeDeleted: [],
+            dataInfoIdToBeDeleted: [],
+            bidIdToBeDeleted: [],
+        };
+
+        let datasetsArray = [];
+        Object.keys(repackedDatasets).forEach((key) => {
+            const dataset = repackedDatasets[key];
+            if (!dataset.foundOffer && !dataset.foundPurchase && !dataset.chosen) {
+                datasetsArray.push({
+                    expiry: dataset.expiry,
+                    datasetId: key,
+                    dataInfoIds: dataset.dataInfoIds,
+                    bidsIds: dataset.bidsIds,
+                });
+            }
+        });
+
+        datasetsArray = datasetsArray
+            .sort((a, b) => a.expiry - b.expiry)
+            .splice(constants.LOW_ESTIMATED_VALUE_DATASETS_PRUNING_BATCH_NUMBER);
+
+        datasetsArray.forEach((dataset) => {
+            idsForPruning.datasetsToBeDeleted.push({
+                datasetId: dataset.datasetId,
+            });
+            idsForPruning.dataInfoIdToBeDeleted.concat(dataset.dataInfoIds);
```

Comment:
concat method does not change the existing arrays, but instead returns a new array, so we need to add the following:
`idsForPruning.dataInfoIdToBeDeleted = idsForPruning.dataInfoIdToBeDeleted.concat(dataset.dataInfoIds);`

The same should be done for the next line as well